File: 164.maximum-gap.js
/*
 * @lc app=leetcode id=164 lang=javascript
 *
 * [164] Maximum Gap
 */

// @lc code=start
/**
 * @param {number[]} nums
 * @return {number}
 */
var maximumGap = function (nums) {
  const mx = Math.max(...nums);
  const mn = Math.min(...nums);
  const interval = (mx - mn) / (nums.length - 1);
  const bucketIndexMappingFunc = (num) => Math.floor((num - mn) / interval);

  let buckets = [];
  nums.forEach((num) => {
    bucketIndex = bucketIndexMappingFunc(num);
    buckets[bucketIndex] = buckets[bucketIndex] ?? [];
    buckets[bucketIndex].push(num);
  });

  buckets = buckets.filter((bucket) => bucket?.length > 0);
  let maxGap = 0;
  let preMax; // max element in previous bucket
  for (let i = 0; i < buckets.length; ++i) {
    if (i > 0) {
      const curMin = Math.min(...buckets[i]); // min element in current bucket
      maxGap = Math.max(maxGap, Math.min(...buckets[i]), curMin - preMax); // according Pigeonhole principle, the gap between two consecutive buckets is the maximum gap
    }
    preMax = Math.max(...buckets[i]);
  }
  return maxGap;
};
// @lc code=end

maximumGap([3, 6, 9, 1]);

File: solutions/164.maximum-gap.rb
# @param {Integer[]} nums
# @return {Integer}
def maximum_gap(nums)
  count = nums.count
  min = nums.min
  max = nums.max

  return 0 if count < 2 || min == max

  size = ((max-min).to_f/(count-1)).ceil
  bucket = ((max-min)/size + 1).times.map{[nil, nil]}

  nums.each do |num|
    b = bucket[(num-min)/size]
    b[0] = [b[0], num].compact.min
    b[1] = [b[1], num].compact.max
  end

  bucket = bucket.select{|b| b[0]}

  (1...bucket.count).map{|i| bucket[i][0]-bucket[i-1][1]}.max
end
